node/src/isolate/external.cpp

```diff
@@ -206,27 +207,71 @@ struct external_t::inner_t :
     std::atomic<id_t> cur_id;
     std::shared_ptr<dispatch<io::context_tag>> signal_dispatch;
     bool prepared;
-    bool connecting;
 
     std::chrono::time_point<std::chrono::system_clock> last_failed_connect_time;
     std::chrono::milliseconds seal_time;
+
     static constexpr std::chrono::milliseconds min_seal_time {1000ul};
     static constexpr std::chrono::milliseconds max_seal_time {1000000ul};
 
     std::vector<std::shared_ptr<spool_load_t>> spool_queue;
     std::vector<std::shared_ptr<spawn_load_t>> spawn_queue;
 
+    std::atomic_flag connecting_ = ATOMIC_FLAG_INIT;
+
+    // All async methods should capture shared_from_this()
+    struct connection_t : std::enable_shared_from_this<connection_t> {
+
+        using async_completion = std::function<void(const std::error_code&, std::unique_ptr<tcp::socket>)>;
+
+        const std::shared_ptr<inner_t> parent_;
+        logging::logger_t& log_;
+
+        std::unique_ptr<tcp::socket> socket_;
+        asio::deadline_timer connect_timer_;
+
+        connection_t(const std::shared_ptr<inner_t>& parent, logging::logger_t &log, asio::io_service& io):
+            parent_(parent),
+            log_(log),
+            socket_(std::make_unique<tcp::socket>(io)),
+            connect_timer_(io)
+        {}
+
+        ~connection_t() {
+            parent_->connecting_.clear();
```

Comment:
Whole `parent` isn't necessary here. And why isn't `connecting_` cleared in `async_completion` callback?